internal/jobobject/iocp.go

```diff
@@ -0,0 +1,83 @@
+package jobobject
+
+import (
+	"context"
+	"sync"
+	"unsafe"
+
+	"github.com/Microsoft/hcsshim/internal/log"
+	"github.com/Microsoft/hcsshim/internal/winapi"
+	"golang.org/x/sys/windows"
+)
+
+var (
+	ioInitOnce sync.Once
+	initIOErr  error
+	// Global iocp handle that will be re-used for every job object
+	ioCompletionPort windows.Handle
+	// Mapping of job handle to channel that that job object owns to be able to receive
+	// notifications.
+	jobMap = make(map[uintptr]chan interface{})
+)
+
+// AllProcessesExited is a type representing a notification that every process in a job has exited.
+type AllProcessesExited struct{}
+
+// ioCompletionProcessor processes completed async IOs forever
+func pollIOCP(ctx context.Context, iocpHandle windows.Handle) {
+	var (
+		overlapped uintptr
+		code       uint32
+		key        uintptr
+	)
+
+	for {
+		err := winapi.GetQueuedCompletionStatus(iocpHandle, &code, &key, (**windows.Overlapped)(unsafe.Pointer(&overlapped)), windows.INFINITE)
+		if err != nil {
+			log.G(ctx).WithError(err).Error("failed during `GetQueuedCompletionStatus`")
+			continue
+		}
+		if ch, ok := jobMap[key]; ok {
+			// Check code and parse out relevant information related to that notification
+			// that we care about. For now all we handle is the message that all processes
+			// in the job have exited.
+			switch code {
+			case winapi.JOB_OBJECT_MSG_ACTIVE_PROCESS_ZERO:
+				// Send the AllProcessesExited notification. We don't remove the map entry here
+				// as for container scenarios for all intents and purposes this is the
+				// end of the containers life, but for possible future usecases for this package
+				// there's nothing stopping more processes from being added to the job and this
+				// being fired once the count reaches zero again. A call to JobObject.Close() will
+				// remove the entry as the handle is invalidated at this point and the job is
+				// finished for good.
+				ch <- AllProcessesExited{}
+			// Other notifications for completeness and a check to make sure that if we fall
+			// into the default case that this is a code we don't know how to handle.
+			case winapi.JOB_OBJECT_MSG_END_OF_JOB_TIME:
+			case winapi.JOB_OBJECT_MSG_END_OF_PROCESS_TIME:
+			case winapi.JOB_OBJECT_MSG_ACTIVE_PROCESS_LIMIT:
+			case winapi.JOB_OBJECT_MSG_NEW_PROCESS:
+			case winapi.JOB_OBJECT_MSG_EXIT_PROCESS:
+			case winapi.JOB_OBJECT_MSG_ABNORMAL_EXIT_PROCESS:
+			case winapi.JOB_OBJECT_MSG_PROCESS_MEMORY_LIMIT:
+			case winapi.JOB_OBJECT_MSG_JOB_MEMORY_LIMIT:
+			case winapi.JOB_OBJECT_MSG_NOTIFICATION_LIMIT:
+			default:
+				log.G(ctx).WithField("code", code).Error("received an unknown iocp code")
```

Comment:
I'd make this warn or info level. There could be new messages added in future releases, and we don't want to get errors logged in those cases.

Reply:
Also don't think we should call this "iocp code" in the log message. The fact it is received using an IOCP isn't very important. Let's just say something like `received job notification of unknown type`.